### PR TITLE
test(charts/chronicle): add chronicle api helm chart test

### PR DESCRIPTION
**Pull request by suchapalaver**
# Instructions

To test this without running a full `make build` change the `image` value in `api-test.yaml` to `dev.catenasys.com:8083/chronicle-helm-api-test:latest`. You can also override this value at install time like this:

```bash
helm install test charts/chronicle/ --set test.image.repository='dev.catenasys.com:8083/chronicle-helm-api-test' --set test.image.tag=latest                
```

The container's `command` value can remain the default.

Create a cluster.

```bash
kind create cluster
```

Create the `regcred` Secret (see the `Job` `container` definition, and also [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account) for instructions on adding `ImagePullSecret`s to a service account). This provides your credentials to access the nexus repository where the docker image for the test is located. The Dockerfile and associated scripts are provided below.

Install the chart.

```bash
helm install <installation-name> chronicle/charts/chronicle
```

Check everything's running.

```bash
kubectl get pods
```

Run the Helm Chart tests

```bash
helm test <installation-name>
```

You should see something like

```
helm test <installation-name>
NAME: <installation-name>
LAST DEPLOYED: Mon May 15 18:34:54 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE:     chronicle-define-agent
Last Started:   Mon May 15 18:35:39 2023
Last Completed: Mon May 15 18:36:04 2023
Phase:          Succeeded
```

Logs for the `test` container, if successful, should contain something like

```
Subscription responses:
{
  "data": {
    "commitNotifications": {
      "stage": "SUBMIT",
      "txId": "b9276c4c6325332011ef9949a60bc3954349fba4ea691b6b38c47f063b6aee874a4d6dcfd8aa2b173c521ae3b4dc694391df90964181373ca6a1ee0bf5c6a9ad"
    }
  }
}
{
  "data": {
    "commitNotifications": {
      "stage": "COMMIT",
      "txId": "b9276c4c6325332011ef9949a60bc3954349fba4ea691b6b38c47f063b6aee874a4d6dcfd8aa2b173c521ae3b4dc694391df90964181373ca6a1ee0bf5c6a9ad"
    }
  }
}
submitted: b9276c4c6325332011ef9949a60bc3954349fba4ea691b6b38c47f063b6aee874a4d6dcfd8aa2b173c521ae3b4dc694391df90964181373ca6a1ee0bf5c6a9ad
committed: b9276c4c6325332011ef9949a60bc3954349fba4ea691b6b38c47f063b6aee874a4d6dcfd8aa2b173c521ae3b4dc694391df90964181373ca6a1ee0bf5c6a9ad
Test passes
```

[CHRON-349](https://blockchaintp.atlassian.net/browse/CHRON-349)

Signed-off-by: Joseph Livesey [joseph.livesey@btp.works](mailto:joseph.livesey@btp.works)

[CHRON-330]: https://blockchaintp.atlassian.net/browse/CHRON-330?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CHRON-349]: https://blockchaintp.atlassian.net/browse/CHRON-349?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: charts/chronicle/templates/tests/define-agent.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            - name: PORT
+              value: {{ .Values.port | quote }}
+          volumeMounts:
+            - name: config
+              mountPath: /scripts
+            - name: shared-volume
+              mountPath: /shared
+      containers:
+        - name: jq
+          image: stedolan/jq
```
Comment (scealiontach) [inline charts/chronicle/templates/tests/define-agent.yaml:27]: use one container image that we define in this project.

File: charts/chronicle/templates/tests/define-agent.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+          command: ["/bin/sh"]
+          args: ["-c", "scripts/compare-txids.sh"]
+          volumeMounts:
+            - name: config
+              mountPath: /scripts
+            - name: shared-volume
+              mountPath: /shared
+      volumes:
+        - name: config
```
Comment (scealiontach) [inline charts/chronicle/templates/tests/define-agent.yaml:36]: use a proper container image and you can toss the config map, and have more of a swiss army knife for testing.

**Review by scealiontach — CHANGES_REQUESTED** [review]

File: charts/chronicle/templates/tests/define-agent.yaml (hunk 1 above)
Comment (suchapalaver) [reply charts/chronicle/templates/tests/define-agent.yaml:27]: Was having issues with dependencies trying to install `jq` in the `node` container. That was part of the idea.

File: charts/chronicle/templates/tests/define-agent.yaml (hunk 2 above)
Comment (suchapalaver) [reply charts/chronicle/templates/tests/define-agent.yaml:36]: What do you mean by "proper"? Would you put the Dockerfile for that in with the Chart? Publish to our image repository?

**Comment by mtbc** [comment]
Needed to wait a long while for initialization so the first attempt timed out (maybe downloading images?), next one was far happier:
```
NAME: pr-302
LAST DEPLOYED: Tue May 16 13:01:25 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE:     chronicle-define-agent
Last Started:   Tue May 16 13:46:19 2023
Last Completed: Tue May 16 13:47:34 2023
Phase:          Succeeded
```

**Review by scealiontach — CHANGES_REQUESTED** [review]

**Review by scealiontach — APPROVED** [review]

File: charts/chronicle/templates/tests/api-test.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        - name: test
+          image: {{ .Values.test.container.image }}
+          command: {{ .Values.test.container.command }}
+          env:
+            - name: HOST
+              value: {{ include "chronicle.service.name" . }}-chronicle-api
+            - name: PORT
+              value: {{ .Values.port | quote }}
+          imagePullPolicy: {{ .Values.test.container.imagePullPolicy }}
+      imagePullSecrets:
```
Comment (scealiontach) [inline charts/chronicle/templates/tests/api-test.yaml:21]: this doesn't belong here.  You should be using the chart service account

**Review by scealiontach — CHANGES_REQUESTED** [review]

**Review by scealiontach — APPROVED** [review]

**Comment by suchapalaver** [comment]
> ```
> mtbc@pop-os:~/src/chronicle$ helm test pr-302
> NAME: pr-302
> LAST DEPLOYED: Thu May 18 08:27:46 2023
> NAMESPACE: chronicle
> STATUS: deployed
> REVISION: 1
> TEST SUITE:     pr-302-chronicle-api-test
> Last Started:   Thu May 18 09:58:56 2023
> Last Completed: Thu May 18 10:03:57 2023
> Phase:          Failed
> Error: 1 error occurred:
> 	* timed out waiting for the condition
> 
> 
> mtbc@pop-os:~/src/chronicle$ helm test pr-302
> NAME: pr-302
> LAST DEPLOYED: Thu May 18 08:27:46 2023
> NAMESPACE: chronicle
> STATUS: deployed
> REVISION: 1
> TEST SUITE:     pr-302-chronicle-api-test
> Last Started:   Thu May 18 10:11:48 2023
> Last Completed: Thu May 18 10:11:48 2023
> Phase:          Failed
> Error: warning: Hook test chronicle/templates/tests/api-test.yaml failed: 1 error occurred:
> 	* jobs.batch "pr-302-chronicle-api-test" already exists
> ```
> 
> Is there a good way to "reset" it so I can try again?

I'm guessing you didn't configure a secret to pull the image or maybe you didn't override default values upon `helm install pr-302` using `--set`? I would try to `kubectl delete job <test>` to reset.

**Comment by mtbc** [comment]
Thank you for the tips. Very plausible that I'm missing some secret configuration. Will discuss on Slack to figure out what I'm doing wrong.

**Pull request merged**